ggu/main.c

```diff
@@ -16,11 +16,295 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <stdbool.h>
+
+#include <rte_ip.h>
+#include <rte_udp.h>
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_malloc.h>
+#include <rte_ethdev.h>
+#include <rte_atomic.h>
+
+#include "gatekeeper_gk.h"
 #include "gatekeeper_ggu.h"
+#include "gatekeeper_net.h"
+#include "gatekeeper_main.h"
+#include "gatekeeper_config.h"
+
+static rte_atomic16_t num_ggu_conf_alloc = RTE_ATOMIC16_INIT(0);
+
+static int
+policy_parse_next(uint16_t *proto_ptr, uint8_t *policy_ptr,
+	struct ggu_policy *policy)
+{
+	if (proto_ptr == NULL || policy_ptr == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"ggu: invalid pointor when parsing policies at %s\n", __func__);
+		return -1;
+	}
+
+	policy->flow.proto = rte_be_to_cpu_16(*(uint16_t *)proto_ptr);
+	if (policy->flow.proto == ETHER_TYPE_IPv4)
+		rte_memcpy(&policy->flow.f.v4, policy_ptr,
+			sizeof(policy->flow.f.v4));
+	else if (policy->flow.proto == ETHER_TYPE_IPv6)
+		rte_memcpy(&policy->flow.f.v6, policy_ptr,
+			sizeof(policy->flow.f.v6));
+	else {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: parse policy with unknown network layer protocol %hhu\n",
+			policy->flow.proto);
+		return -1;
+	}
+
+	rte_memcpy(&policy->params, policy_ptr, sizeof(policy->params));
+
+	return 0;
+}
+
+static void
+process_single_policy(struct ggu_policy *policy, struct ggu_config *ggu_conf)
+{
+	/*
+	 * Identify which GK block is responsible for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	int block_id = get_responsible_block(&policy->flow, ggu_conf->gk);
+	struct gk_cmd_entry entry;
+
+	if (block_id < 0)
```

Comment:
No, the warning in `get_responsible_block()` is enough. I had missed that. Thank you.